core/services/ocr2/plugins/ocr2vrf/fixed_blockhash_provider.go

```diff
@@ -0,0 +1,109 @@
+package ocr2vrf
+
+import (
+	"context"
+	"math/big"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/smartcontractkit/ocr2vrf/types"
+
+	"github.com/smartcontractkit/chainlink/core/chains/evm/client"
+)
+
+// Implementation of Blockhashes interface that returns blockhashes with fixed-sized windows relative to current block height
+type FixedBlockhashProvider struct {
+	client client.Client
+	// end block = current head - waitBlocks
+	waitBlocks uint
+	// start block = current head - lookbackBlocks
+	lookbackBlocks uint
+	// number of blocks to query in a batch
+	batchSize uint
+}
+
+type Block struct {
+	Number    int64
+	Hash      common.Hash
+	Timestamp time.Time
+}
```

Comment:
Is this necessary or can we just re-use go-ethereum's [`types.Header`](https://github.com/ethereum/go-ethereum/blob/master/core/types/block.go#L70)? I would lean towards using `types.Header` which reduces duplication and maintenance overhead.

Reply:
Hmm, seems like `types.Header` has different fields than what's expected from the RPC call. We can use [evmtypes.Head](https://github.com/smartcontractkit/chainlink/blob/develop/core/chains/evm/types/models.go#L22) to avoid duplication.

Reply:
Interesting, I'm pretty shocked it doesn't work. Here's the schema for `eth_getBlockByNumber`: https://ethereum.github.io/execution-apis/api-documentation/. Glancing over the fields, it seems like `types.Header` should implement it (the `json` struct tags match 1-1 with the returned object). In contrast, `evmtypes.Head` seems pretty different, it looks like what we'd store in the chainlink PG database, in the `evm_heads` table.

Reply:
Here's a code snippet that works for me:

```golang
tip, err := e.Ec.HeaderByNumber(context.Background(), nil)
		helpers.PanicErr(err)
		var reqs []rpc.BatchElem
		for i := 1; i <= 10; i++ {
			reqs = append(reqs, rpc.BatchElem{
				Method: "eth_getBlockByNumber",
				Args: []any{
					hexutil.EncodeBig(new(big.Int).Sub(tip.Number, big.NewInt(int64(i)))),
					true,
				},
				Result: &types.Header{},
			})
		}

		err = e.JRPC.BatchCallContext(context.Background(), reqs)
		helpers.PanicErr(err)

		for _, r := range reqs {
			helpers.PanicErr(r.Error)
			header, ok := r.Result.(*types.Header)
			if !ok {
				fmt.Println("not of type header")
				os.Exit(1)
			}
			if header == nil {
				fmt.Println("nil block")
				os.Exit(1)
			}
			if header.Hash() == (common.Hash{}) {
				fmt.Println("nil hash")
				os.Exit(1)
			}
			fmt.Println("block hash of block", r.Args[0], "is", header.Hash())
		}
```

Reply:
Thanks for the snippet! I agree that the geth types are preferred model for RPC calls over our own structs. That being said, I still have one open question:

The comment [here](https://github.com/smartcontractkit/chainlink/blob/develop/core/chains/evm/client/client.go#L50-L54) says that we are using our own types intentionally to capture to correct hash for Parity nodes on Kovan.

Reply:
I think that would be useful (to be able to request from both Parity and go-ethereum and other providers), but using `&evmtypes.Head{}` in place of `&types.Header{}` simply doesn't work for me. This is the error I get:

```
not of type header
```

which is thrown when trying to unmarshal the returned JSON-RPC response of `eth_getBlockByNumber`. This makes sense because `evmtypes.Head` doesn't actually have the necessary fields to unmarshal into.

Reply:
Looping back to this, `evmtypes.Head` seems to work now. I must've fudged the import originally. All good!